cvmfs/sync_mediator.cc

```diff
@@ -920,9 +920,16 @@ void SyncMediator::AddDirectory(SharedPtr<SyncItem> entry) {
   }
 
   if (entry->HasCatalogMarker() &&
-      !catalog_manager_->IsTransitionPoint(entry->GetRelativePath())) {
+    !catalog_manager_->IsTransitionPoint(entry->GetRelativePath())) {
     CreateNestedCatalog(entry);
-  }
+    } else if (!entry->HasCatalogMarker()) {
+      if (entry->IsPathPresent()) {
```

Comment:
This logic is not quite correct: we are in `AddDirectory`, so `entry` represents a directory.  Now we go into the `else` branch, so the directory `entry` does _not_ contain a .cvmfscatalog regular file.  Then we test if the directory itself is present (it should) and in the next line if it is a regular file (it isn't, we know already that it is a directory).

So this block would panic whenever we try to add a directory that does not contain a .cvmfscatalog regular file.

Let's not put the changes here, directly, but let's instead patch `SyncItem::CheckCatalogMarker()`